src/peergos/server/tests/simulation/Simulator.java

```diff
@@ -136,63 +159,172 @@ private int getNextFileLength() {
+
         byte[] refBytes = referenceFileSystem.read(path);
         byte[] testBytes = testFileSystem.read(path);
         return Arrays.equals(refBytes, testBytes);
     }
 
-    private void write() {
-        Path path = getAvailableFilePath();
+    private void write(String user) {
+        Path path = getAvailableFilePath(user);
         byte[] fileContents = getNextFileContents();
 
         Path dirName = path.getParent();
         String fileName = path.getFileName().toString();
-        simulatedDirectoryToFiles.get(dirName).add(fileName);
+        Map<Path, List<String>> dirsToFiles = index.getDirToFiles(user);
+        dirsToFiles.get(dirName).add(fileName);
 
         LOG.info("Writing path " + path.resolve(fileName) + " with length " + fileContents.length);
 
+        FileSystem testFileSystem = fileSystems.getTestFileSystem(user);
+        FileSystem referenceFileSystem = fileSystems.getReferenceFileSystem(user);
+
         testFileSystem.write(path, fileContents);
         referenceFileSystem.write(path, fileContents);
     }
 
+    private boolean grantRead(String granter, String grantee, Path path) {
+        LOG.info("Granting read-access to " + path + " for granter " + granter + " and grantee " + grantee);
+
+
+        FileSystem testFileSystem = fileSystems.getTestFileSystem(granter);
+        FileSystem referenceFileSystem = fileSystems.getReferenceFileSystem(granter);
+
+        testFileSystem.grant(path, grantee, FileSystem.Permission.READ);
+        referenceFileSystem.grant(path, grantee, FileSystem.Permission.READ);
+
+        return true;
+    }
+
+    private boolean revokeRead(String revoker, String revokee, Path path) {
+        LOG.info("Revoking read-access to " + path + "for revoker " + revoker + " and revokee " + revokee);
+
+
+        FileSystem testFileSystem = fileSystems.getTestFileSystem(revoker);
+        FileSystem referenceFileSystem = fileSystems.getReferenceFileSystem(revoker);
+
+        testFileSystem.revoke(path, revokee, FileSystem.Permission.READ);
+        referenceFileSystem.revoke(path, revokee, FileSystem.Permission.READ);
+
+        return true;
+    }
+
     private void init() {
-        //seed file-system with a directory and a file
-        run(Simulation.MKDIR);
-        run(Simulation.WRITE);
+        List<String> users = this.fileSystems.getUsers();
+        Collections.sort(users);
+
+        for (String user : users) {
+            index.addUser(user);
+            index.getDirToFiles(user).put(Paths.get("/" + user), new ArrayList<>());
+            // seed the file-system
+            run(Simulation.MKDIR, user);
+            run(Simulation.WRITE, user);
+        }
     }
 
     private void run(Simulation simulation) {
+        String nextUser = fileSystems.getNextUser();
+        run(simulation, nextUser);
+    }
+
+    private void run(Simulation simulation, String user) {
+        Supplier<String> otherUser = () -> fileSystems.getNextUser(user);
```

Comment:
I imagine it's going to be super useful to have a single line log with op : user [path]